### PR TITLE
Time based validation support

### LINKED ISSUE #13324: Validation takes place every N time
## 🚀 Feature

validation takes place every N time (ex. validation takes place every 15 minutes )

### Motivation
I would like to be able to run validation once at some fixed time.
I know that validation takes a specific time, and in order not to spend more training time on validation, I want to be able to run training once in a specific period of time.

For example, my validation takes 5 minutes (on the CPU, because this is RL). then I want to run it once every 15, so that the time spent on validation is no more than 25% of the total training time.
This is due to the fact that on different machines and servers, learning has a different speed, and if we start from the number of iterations, then it may be that we either have too frequent validation and most of the time is spent on it, or too rare, and we could do it more often if we were guided by the elapsed time from the previous validation.

Yes, I understand that this is not a universal solution, but it is better than counting the number of iterations, as we do now)

 When changing the network architecture or even selecting, the time changes iterations begins to change in a more complicated way.

the training itself takes place on the GPU

### Pitch

so that you can set a time interval, as is done with max_time , once in which validation will be called. Alternatively, this can be tied to the end of the epoch, roughly speaking, if the timer has come, then as soon as the epoch is over, then we go to do the validation. Or even make an additional flag that would say to do validation when time expires or when time expires and the end of the epoch.

### Alternatives

described above

### Additional context

slack link https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1655476166180589

______________________________________________________________________

Thank you very much for the lightning, we love and use it very much ❤️

cc @borda @justusschock @kaushikb11 @awaelchli @rohitgr7 @ninginthecloud

### PR DESCRIPTION
**Pull request by Sohaib-Ahmed21**
## What does this PR do?

- Adds an option to run validation **by time** (e.g., every 15 seconds) via `val_check_interval` (string, `timedelta`, or dict). 
- The timer starts at fit, triggers validation when elapsed, respects `check_val_every_n_epoch`, and resets after each validation.
-  In time mode we skip epoch-end checkpoints unless `save_on_train_epoch_end` is set to `True`.
-  Existing int/float behavior stays the same.
-  Tests updated.

Fixes #13324 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
yes
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
Will do, it's draft PR now.
- Did you write any **new necessary tests**? (not for typos and docs)
- Yes
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
No breaking changes introduced
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)
Will update, it's draft now.

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by Borda — APPROVED** [review]
I like the extension and I think it valuable use-case for continues training, but leaving the decision to @lantiga

**Review by lantiga — APPROVED** [review]
Thank you for the contribution @Sohaib-Ahmed21

Goes without saying, runs become non reproducible, as in:
- it's not possible to re-run and have validation happen at the same intervals
- interrupting a run and resuming will not lead to the same validation curves

I would make sure users understand this, so in case they use validation loss as a criterion for scheduling or stopping they will be fully aware.

File: src/lightning/pytorch/trainer/setup.py [hunk 1]
```diff
@@ -187,3 +196,31 @@ def _log_device_info(trainer: "pl.Trainer") -> None:
 
         if HPUAccelerator.is_available() and not isinstance(trainer.accelerator, HPUAccelerator):
             rank_zero_warn("HPU available but not used. You can set it by doing `Trainer(accelerator='hpu')`.")
+
+
+def _parse_time_interval_seconds(value: Union[str, timedelta, dict]) -> float:
```
Comment (SkafteNicki) [inline src/lightning/pytorch/trainer/setup.py:201]:
```suggestion
def _parse_time_interval_seconds(value: Union[str, timedelta, dict]) -> float:
    """Convert a time interval into seconds.

    This helper parses different representations of a time interval and
    normalizes them into a float number of seconds.

    Supported input formats:
      * `timedelta`: The total seconds are returned directly.
      * `dict`: A dictionary of keyword arguments accepted by
        `datetime.timedelta`, e.g. `{"days": 1, "hours": 2}`.
      * `str`: A string in the format `"DD:HH:MM:SS"`, where each
        component must be an integer.

    Args:
        value (Union[str, timedelta, dict]): The time interval to parse.

    Returns:
        float: The duration represented by `value` in seconds.

    Raises:
        MisconfigurationException: If the input type is unsupported, the
        string format is invalid, or any string component is not an integer.

    Examples:
        >>> _parse_time_interval_seconds("01:02:03:04")
        93784.0

        >>> _parse_time_interval_seconds({"hours": 2, "minutes": 30})
        9000.0

        >>> from datetime import timedelta
        >>> _parse_time_interval_seconds(timedelta(days=1, seconds=30))
        86430.0
    """
```

File: tests/tests_pytorch/trainer/flags/test_val_check_interval.py [hunk 2]
```diff
@@ -127,9 +130,110 @@ def test_val_check_interval_float_with_none_check_val_every_n_epoch():
...
+        match=re.escape(
+            "`val_check_interval` should be an integer or a time-based duration (str 'DD:HH:MM:SS', "
+            "datetime.timedelta, or dict kwargs for timedelta) when `check_val_every_n_epoch=None`."
+        ),
     ):
         Trainer(
             val_check_interval=0.5,
             check_val_every_n_epoch=None,
         )
+
```
Comment (SkafteNicki) [inline tests/tests_pytorch/trainer/flags/test_val_check_interval.py:144]: it seems neither of these tests are checking that the time based validation works for all 3 input formats (string, dict, timedelta). Consider adding a test for `_parse_time_interval_seconds` or to parameterize one of the tests for the different formats

**Review by Borda — APPROVED** [review]
